File: studio/backend/utils/models/hidden_models.py
# SPDX-License-Identifier: AGPL-3.0-only
# Copyright 2026-present the Unsloth AI Inc. team. All rights reserved. See /studio/LICENSE.AGPL-3.0

"""Infra-only model detection shared by the model routes and the hub
inventory. Lives in utils so hub services can use it without importing
routes.models (which has import-time side effects and would cycle)."""

from __future__ import annotations

import re
from pathlib import Path
from typing import Optional

# Hub repo id shape ("owner/name", no leading separator); anything else is
# treated as a local filesystem path.
_HF_REPO_ID_RE = re.compile(r"^[A-Za-z0-9][\w.\-]*/[\w.\-]+$")


def _safe_resolve(path: Path) -> Optional[str]:
    """resolve() to a string, or None when the path is inaccessible."""
    try:
        return str(path.resolve())
    except OSError:
        return None


def is_hidden_model(*values: str | None) -> bool:
    """True if any id/path is the RAG embedding model (EMBEDDING_MODEL or
    EMBED_GGUF_REPO basename) or the llama.cpp install validation probe
    (ggml-org/models / stories260K), so pickers hide them (GGUF and non-GGUF).
    None are usable chat models; the probe can be cached as a side effect of
    installing the prebuilt llama-server and otherwise sorts smallest, so it
    would be auto-selected. A local-path embedder is matched by exact resolved
    path only: a generic basename like "model" must not substring-hide
    unrelated chat models."""
    from core.rag import config as rag_config

    needles = [
        # The validation probe's repo (matches the cached repo id) and its exact
        # filename (matches the on-disk path). The filename carries the .gguf so
        # it does not hide unrelated repos like ``user/stories260K-finetune-GGUF``.
        "ggml-org/models",
        "stories260k.gguf",
    ]
    exact_paths: list[str] = []
    for model in (
        rag_config.effective_embedding_model(),
        rag_config.effective_gguf_repo(),
    ):
        if _HF_REPO_ID_RE.match(model):
            needles.append(model.split("/")[-1].lower())
        else:
            resolved = _safe_resolve(Path(model).expanduser())
            if resolved:
                exact_paths.append(resolved.lower())
    for v in values:
        if not v:
            continue
        low = v.lower()
        if any(n in low for n in needles):
            return True
        if exact_paths:
            resolved = _safe_resolve(Path(v).expanduser())
            if resolved and resolved.lower() in exact_paths:
                return True
    return False

File: studio/frontend/src/features/hub/lib/hidden-models.ts
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright 2026-present the Unsloth AI Inc. team. All rights reserved. See /studio/LICENSE.AGPL-3.0

import { authFetch } from "@/features/auth";
import { getInventoryVersion } from "../stores/inventory-events";
import {
  getHiddenModelMatchersSnapshot,
  type HiddenModelMatchersSnapshot,
  replaceHiddenModelMatchers,
} from "./hidden-model-matcher-store";

// Infra models hidden from browse/preview lists (Hub Discover, the chat model
// selector, and local on-device rows). Mirrors the backend
// `utils.hidden_models`: the RAG embedding model, STT dictation models, and the
// llama.cpp validation probe are not usable chat models. Server-confirmed cache
// rows are trusted because the backend applies variant-aware filtering.
// Optimistic cache rows still use these needles until the server confirms them.
// The dynamic matchers fetched from `/api/hub/hidden-models` add default and
// configured embedders as exact repo ids and exact resolved paths, never
// substring needles. Per-repo views are not filtered, so reinstall flows still
// show downloaded files.
const HIDDEN_NEEDLES = [
  "bge-small-en-v1.5", // RAG embedder: unsloth/bge-small-en-v1.5[-GGUF]
  "ggml-org/models", // llama.cpp validation probe repo
  "stories260k.gguf", // probe filename (carries .gguf so it stays specific)
];
const HIDDEN_STT_REPOS = new Set([
  // Transformers safetensors repos and their whisper.cpp GGUF companions
  // (unslothai/whisper-*-GGUF): STT-only, never chat models. The Qwen3-ASR
  // GGUFs are here for the same reason: llama.cpp will load one as a chat
  // model, where it only answers with transcripts.
  "unsloth/whisper-tiny",
  "unsloth/whisper-base",
  "unsloth/whisper-small",
  "unsloth/whisper-large-v3-turbo",
  "unsloth/whisper-large-v3",
  "unslothai/whisper-tiny-gguf",
  "unslothai/whisper-base-gguf",
  "unslothai/whisper-small-gguf",
  "unslothai/whisper-large-v3-turbo-gguf",
  "unslothai/whisper-large-v3-gguf",
  "unslothai/qwen3-asr-0.6b-gguf",
  "unslothai/qwen3-asr-1.7b-gguf",
]);
const HIDDEN_STT_CACHE_NAMES = [...HIDDEN_STT_REPOS].map((repo) =>
  repo.replace("/", "--"),
);

let matchersFetch: Promise<void> | null = null;
let matchersFetchVersion = -1;

function toLowerStrings(value: unknown): string[] {
  if (!Array.isArray(value)) {
    return [];
  }
  return value
    .filter((v): v is string => typeof v === "string" && v.length > 0)
    .map((v) => v.toLowerCase());
}

export function ensureHiddenModelMatchers(): Promise<void> {
  const version = getInventoryVersion();
  if (matchersFetch && matchersFetchVersion === version) {
    return matchersFetch;
  }
  matchersFetchVersion = version;
  matchersFetch = (async () => {
    try {
      const response = await authFetch("/api/hub/hidden-models");
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }
      const data = (await response.json()) as {
        needles?: unknown;
        exact_ids?: unknown;
        exact_paths?: unknown;
      };
      if (
        getInventoryVersion() !== version ||
        matchersFetchVersion !== version
      ) {
        return;
      }
      replaceHiddenModelMatchers({
        needles: toLowerStrings(data.needles),
        exactIds: toLowerStrings(data.exact_ids),
        exactPaths: toLowerStrings(data.exact_paths),
      });
    } catch {
      if (
        getInventoryVersion() === version &&
        matchersFetchVersion === version
      ) {
        matchersFetch = null;
      }
    }
  })();
  return matchersFetch;
}

/** True if any id/path is a hidden infra model. */
export function isHiddenModelIdWithMatchers(
  matchers: HiddenModelMatchersSnapshot,
  ...values: (string | null | undefined)[]
): boolean {
  return values.some((v) => {
    if (!v) {
      return false;
    }
    const lower = v.toLowerCase();
    const normalized = lower.trim().replace(/^\/+|\/+$/g, "");
    const pathParts = lower.split(/[\\/]/);
    return (
      HIDDEN_STT_REPOS.has(normalized) ||
      HIDDEN_STT_CACHE_NAMES.some((name) =>
        pathParts.includes(`models--${name}`),
      ) ||
      HIDDEN_NEEDLES.some((needle) => lower.includes(needle)) ||
      matchers.needles.some((needle) => lower.includes(needle)) ||
      matchers.exactIds.includes(lower) ||
      matchers.exactPaths.includes(lower)
    );
  });
}

export function isHiddenModelId(
  ...values: (string | null | undefined)[]
): boolean {
  return isHiddenModelIdWithMatchers(
    getHiddenModelMatchersSnapshot(),
    ...values,
  );
}
